common/state/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+}
+
+impl FileOffsetFlushHandle {
+    pub async fn flush(&self) -> Result<(), FileOffsetStateError> {
+        Ok(self.tx.send(FileOffsetEvent::Flush).await?)
+    }
+
+    pub async fn clear(&self) -> Result<(), FileOffsetStateError> {
+        Ok(self.tx.send(FileOffsetEvent::Clear).await?)
+    }
+}
+
+pub struct FileOffsetShutdownHandle {
+    tx: async_channel::Sender<FileOffsetEvent>,
+}
+
+impl FileOffsetShutdownHandle {
+    pub fn shutdown(&self) -> bool {
+        self.tx.close()
+    }
+}
+
+#[derive(Clone)]
+pub struct FileOffsetState {
+    db: Arc<DB>,
+    cf_opts: Options,
+    rx: std::cell::RefCell<Option<async_channel::Receiver<FileOffsetEvent>>>,
+    shutdown: std::cell::RefCell<Option<async_channel::Sender<FileOffsetEvent>>>,
+    tx: async_channel::Sender<FileOffsetEvent>,
+}
+
+impl FileOffsetState {
+    fn new(db: Arc<DB>, cf_opts: Options) -> Self {
+        let (tx, rx) = async_channel::unbounded();
+
+        FileOffsetState {
+            db,
+            cf_opts,
+            rx: std::cell::RefCell::new(Some(rx)),
+            shutdown: std::cell::RefCell::new(Some(tx.clone())),
+            tx,
+        }
+    }
+
+    pub fn offsets(&self) -> Vec<FileOffset> {
+        let cf_handle = self.db.cf_handle(OFFSET_NAME).unwrap();
+        self.db
+            .iterator_cf(cf_handle, IteratorMode::Start)
+            .map(|(k, v)| {
+                let (int_bytes, _) = v.split_at(std::mem::size_of::<u64>());
+                FileOffset {
+                    key: FileName(bytes::Bytes::copy_from_slice(k.as_ref())),
+                    offset: u64::from_be_bytes(int_bytes.try_into().unwrap_or([0; 8])),
+                }
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub fn write_handle(&self) -> FileOffsetWriteHandle {
+        FileOffsetWriteHandle {
+            tx: self.tx.clone(),
+        }
+    }
+
+    pub fn flush_handle(&self) -> FileOffsetFlushHandle {
+        FileOffsetFlushHandle {
+            tx: self.tx.clone(),
+        }
+    }
+
+    pub fn shutdown_handle(&self) -> Result<FileOffsetShutdownHandle, FileOffsetStateError> {
+        Ok(FileOffsetShutdownHandle {
+            tx: self
+                .shutdown
+                .borrow_mut()
+                .take()
+                .ok_or(FileOffsetStateError::ShutdownHandleTaken)?,
+        })
+    }
+
+    pub fn run(&self) -> Result<impl std::future::Future<Output = ()>, FileOffsetStateError> {
+        let rx = self
+            .rx
+            .borrow_mut()
+            .take()
+            .ok_or(FileOffsetStateError::AlreadyRunning)?;
+        let self_ref = self.db.clone();
+        Ok(rx
+            .fold(Some(WriteBatch::default()), {
+                let self_ref = self_ref.clone();
+                move |a, e| {
+                    let self_ref = self_ref.clone();
+                    async move {
+                        let cf_handle = self_ref.cf_handle(OFFSET_NAME).unwrap();
+                        match (a, e) {
```

Comment:
Nice db grouping of operations and the pattern matching groups, simplifies the logic quite a bit!